File: 백준/Gold/1052. 물병/물병.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        PriorityQueue<Integer> pq = new PriorityQueue<>();
        int num = 0;
        int mul = N;
        while (mul != 0) {
            int mod = mul % 2;
            mul /= 2;
            if (mod != 0) pq.offer((int) Math.pow(2, num));
            num++;
        }

        int result = 0;
        while (pq.size() > K) {
            int cur = pq.poll();
            int next = pq.poll();
            result += next - cur;
            pq.offer(next * 2);
        }
        System.out.println(result);
    }
}

File: 백준/Gold/1052. 물병/물병.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    var st = StringTokenizer(readLine())
    var N = st.nextToken().toInt()
    var K = st.nextToken().toInt()

    var pq = PriorityQueue<Int>()
    var mul = N;
    var num = 0;
    while (mul != 0) {
        val mod = mul % 2
        mul /= 2
        if (mod != 0) pq.offer(Math.pow(2.0, num.toDouble()).toInt())
        num++
    }

    var result = 0
    while (pq.size > K) {
        val cur = pq.poll()
        val next = pq.poll()
        result += next - cur
        pq.offer(next * 2)
    }
    println(result)
}